test/Test/Database/LSMTree/Internal/Entry.hs

```diff
@@ -25,96 +28,352 @@ tests :: TestTree
       (e1@Monoidal.Insert{}, _                  ) -> e1
       (Monoidal.Mupsert v1 , Monoidal.Delete    ) -> Monoidal.Insert v1
       (Monoidal.Mupsert v1 , Monoidal.Insert  v2) -> Monoidal.Insert (v1 <> v2)
       (Monoidal.Mupsert v1 , Monoidal.Mupsert v2) -> Monoidal.Mupsert (v1 <> v2)
 
-newtype EntrySG v b = EntrySG (Entry v b)
+--
+-- Entry
+--
+
+deriving via Uniform (Entry v b)
+    instance (Arbitrary v, Arbitrary b) => Arbitrary (Regular (Entry v b))
+
+-- | Semigroup instance using 'combine'.
+instance Semigroup v => Semigroup (Regular (Entry v b)) where
+  Regular e1 <> Regular e2 = Regular $ combine (<>) e1 e2
+
+{-------------------------------------------------------------------------------
+  Union semantics
+-------------------------------------------------------------------------------}
+
+newtype Union a = Union { unUnion :: a }
+  deriving stock (Show, Eq)
+
+--
+-- Unified update
+--
+
+deriving newtype instance (Arbitrary v, Arbitrary b)
+                       => Arbitrary (Union (Unified.Update v b))
+
+instance Semigroup v => Semigroup (Union (Unified.Update v b)) where
+  Union up1 <> Union up2 = Union $ case (up1, up2) of
+      (Unified.Delete       , _                    ) -> up2
+      (_                    , Unified.Delete       ) -> up1
+      (Unified.Insert v1 mb1, Unified.Insert v2 mb2) ->
+          Unified.Insert
+            (v1 <> v2)
+            (Monoid.getFirst (Monoid.First mb1 <> Monoid.First mb2))
+      (Unified.Insert v1 mb1, Unified.Mupsert v2   ) -> Unified.Insert (v1 <> v2) mb1
+      (Unified.Mupsert v1   , Unified.Insert v2 mb2) -> Unified.Insert (v1 <> v2) mb2
+      (Unified.Mupsert v1   , Unified.Mupsert v2   ) -> Unified.Insert (v1 <> v2) Nothing
+
+--
+-- Normal update
+--
+
+deriving newtype instance (Arbitrary v, Arbitrary b)
+                       => Arbitrary (Union (Normal.Update v b))
+
+instance Semigroup v => Semigroup (Union (Normal.Update v b)) where
+  Union up1 <> Union up2 = Union $ case (up1, up2) of
+    (Normal.Delete       , _                    ) -> up2
+    (_                   , Normal.Delete        ) -> up1
+    (Normal.Insert v1 mb1, Normal.Insert v2 mb2 ) ->
+        Normal.Insert
+          (v1 <> v2)
+          (Monoid.getFirst (Monoid.First mb1 <> Monoid.First mb2))
+
+--
+-- Monoidal update
+--
+
+deriving newtype instance Arbitrary v
+                       => Arbitrary (Union (Monoidal.Update v))
+
+instance Semigroup v => Semigroup (Union (Monoidal.Update v)) where
+  Union up1 <> Union up2 = Union $ case (up1, up2) of
+      (Monoidal.Delete     , _                  ) -> up2
+      (_                   , Monoidal.Delete    ) -> up1
+      (Monoidal.Insert v1  , Monoidal.Insert v2 ) -> Monoidal.Insert (v1 <> v2)
+      (Monoidal.Insert v1  , Monoidal.Mupsert v2) -> Monoidal.Insert (v1 <> v2)
+      (Monoidal.Mupsert v1 , Monoidal.Insert v2 ) -> Monoidal.Insert (v1 <> v2)
+      (Monoidal.Mupsert v1 , Monoidal.Mupsert v2) -> Monoidal.Insert (v1 <> v2)
```

Comment:
I think the semantics we want to compare `combineUnion` to might not be this one on `{Normal,Monoidal}.Update` (which almost copies what `combineUnion` itself does), but the `Semigroup` instance of `Maybe v` and `Maybe (v, Monoid.First b)` instead. I'd say that makes the intended behaviour clearer.